File: Pedometer/ViewController.m
//
//  ViewController.m
//  Pedometer
//
//  Created by Jay Versluis on 29/10/2015.
//  Copyright © 2015 Pinkstone Pictures LLC. All rights reserved.
//

#import "ViewController.h"
@import CoreMotion;

@interface ViewController ()
@property (strong, nonatomic) CMPedometer *pedometer;
@property (strong, nonatomic) IBOutlet UILabel *stepsLabel;
@property (strong, nonatomic) IBOutlet UILabel *distanceLabel;
@property (strong, nonatomic) IBOutlet UILabel *cadenceLabel;
@property (strong, nonatomic) IBOutlet UILabel *paceLabel;
@property (strong, nonatomic) IBOutlet UILabel *flightsUpLabel;
@property (strong, nonatomic) IBOutlet UILabel *flightsDownLabel;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self resetLabels];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (CMPedometer *)pedometer {
    
    if (!_pedometer) {
        _pedometer = [[CMPedometer alloc]init];
    }
    return _pedometer;
}

- (IBAction)startTracking:(id)sender {
    
    self.stepsLabel.text = @"Gathering data...";
    
    // start live tracking
    [self.pedometer startPedometerUpdatesFromDate:[NSDate date] withHandler:^(CMPedometerData * _Nullable pedometerData, NSError * _Nullable error) {
        
        // this block is called for each live update
        [self updateLabels:pedometerData];
        
    }];
}

- (IBAction)stopTracking:(id)sender {
    
    // stop live tracking
    [self.pedometer stopPedometerUpdates];
    [self resetLabels];
}

- (void)resetLabels {
    
    // reset labels
    self.stepsLabel.text = @"Press Start to begin\nPedometer Tracking";
    self.distanceLabel.text = nil;
    self.cadenceLabel.text = nil;
    self.paceLabel.text = nil;
    self.flightsUpLabel.text = nil;
    self.flightsDownLabel.text = nil;
}

- (void)updateLabels:(CMPedometerData *)pedometerData {
    
    NSNumberFormatter *formatter = [[NSNumberFormatter alloc]init];
    formatter.maximumFractionDigits = 2;
    
    // step counting
    if ([CMPedometer isStepCountingAvailable]) {
        self.stepsLabel.text = [NSString stringWithFormat:@"Steps walked: %@", [formatter stringFromNumber:pedometerData.numberOfSteps]];
    } else {
        self.stepsLabel.text = @"Step Counter not available.";
    }
    
    // distance
    if ([CMPedometer isDistanceAvailable]) {
        self.distanceLabel.text = [NSString stringWithFormat:@"Distance travelled: \n%@ meters", [formatter stringFromNumber:pedometerData.distance]];
    } else {
        self.distanceLabel.text = @"Distance estimate not available.";
    }
    
    // pace
    if ([CMPedometer isPaceAvailable] && pedometerData.currentPace) {
        self.paceLabel.text = [NSString stringWithFormat:@"Current Pace: \n%@ seconds per meter", [formatter stringFromNumber:pedometerData.currentPace]];
    } else {
        self.paceLabel.text = @"Pace not available.";
    }
    
    // cadence
    if ([CMPedometer isCadenceAvailable] && pedometerData.currentCadence) {
        self.cadenceLabel.text = [NSString stringWithFormat:@"Cadence: \n%@ steps per second", [formatter stringFromNumber: pedometerData.currentCadence]];
    } else {
        self.cadenceLabel.text = @"Cadence not available.";
    }
    
    // flights climbed
    if ([CMPedometer isFloorCountingAvailable] && pedometerData.floorsAscended) {
        self.flightsUpLabel.text = [NSString stringWithFormat:@"Floors ascended: %@", pedometerData.floorsAscended];
    } else {
        self.flightsUpLabel.text = @"Floors ascended\nnot available.";
    }
    
    if ([CMPedometer isFloorCountingAvailable] && pedometerData.floorsDescended) {
        self.flightsDownLabel.text =[NSString stringWithFormat:@"Floors descended: %@", pedometerData.floorsDescended];
    } else {
        self.flightsDownLabel.text = @"Floors descended\nnot available.";
    }

}












@end

File: Pedometer-Swift/Pedometer-Swift/ViewController.swift
//
//  ViewController.swift
//  Pedometer-Swift
//
//  Created by Jay Versluis on 29/07/2016.
//  Copyright © 2016 Pinkstone Pictures LLC. All rights reserved.
//

import UIKit
import CoreMotion

class ViewController: UIViewController {
    
    @IBOutlet var stepsLabel: UILabel!
    @IBOutlet var distanceLabel: UILabel!
    @IBOutlet var candenceLabel: UILabel!
    @IBOutlet var paceLabel: UILabel!
    @IBOutlet var flightsUpLabel: UILabel!
    @IBOutlet var flightsDownLabel: UILabel!
    
    lazy var pedometer: CMPedometer = {
        let pedometer = CMPedometer.init()
        return pedometer
    }()
    

    override func viewDidLoad() {
        super.viewDidLoad()
        self.resetLabels()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    @IBAction func startTracking(sender: AnyObject) {
        
        self.stepsLabel.text = "Gathering data..."
        
        // start live tracking
        self.pedometer.startPedometerUpdatesFromDate(NSDate()) { (pedometerData: CMPedometerData?, error: NSError?) in
            
            // this block is caleld for each live update
            self.updateLabels(pedometerData!)
        }
    }

    @IBAction func stopTracking(sender: AnyObject) {
        
        self.pedometer.stopPedometerUpdates()
        self.resetLabels()
    }

    func resetLabels() {
        
        self.stepsLabel.text = "Press Start to begin/nPedometer Tracking"
        self.distanceLabel.text = nil
        self.candenceLabel.text = nil
        self.paceLabel.text = nil
        self.flightsUpLabel.text = nil
        self.flightsDownLabel.text = nil
        
    }
    
    func updateLabels(pedometerData: CMPedometerData)  {
        
        // initialise a number formatter
        let formatter = NSNumberFormatter.init()
        formatter.maximumFractionDigits = 2
        
        // step counting
        if CMPedometer.isStepCountingAvailable() {
            self.stepsLabel.text = "Steps walked: \(formatter.stringFromNumber(pedometerData.numberOfSteps))"
        } else {
            self.stepsLabel.text = "Step Counter not available."
        }
        
        // distance
        if CMPedometer.isDistanceAvailable() {
            self.distanceLabel.text = "Distance travelled:\n \(formatter.stringFromNumber(pedometerData.distance!)) meters"
        } else {
            self.distanceLabel.text = "Distance estimate not available."
        }
        
        // pace
        if CMPedometer.isPaceAvailable() {
            self.paceLabel.text = "Current Pace: \n\(formatter.stringFromNumber(pedometerData.currentPace!)) seconds per meter"
        } else {
            self.paceLabel.text = "Pace not available."
        }
        
        // cadence
        if CMPedometer.isCadenceAvailable() {
            self.candenceLabel.text = "Cadence: \n \(formatter.stringFromNumber(pedometerData.currentCadence!)) steps per second"
        } else {
            self.candenceLabel.text = "Cadence not available."
        }
        
        // flights climbed
        if CMPedometer.isFloorCountingAvailable() {
            self.flightsUpLabel.text = "Floors ascended: \(formatter.stringFromNumber(pedometerData.floorsAscended!))"
            self.flightsDownLabel.text = "Floors decended: \(formatter.stringFromNumber(pedometerData.floorsDescended!))"
        } else {
            self.flightsUpLabel.text = "Floor counting not available."
            self.flightsDownLabel.text = nil
        }
    }

}

